### PR TITLE
Change dangerous default random seed selection

### LINKED ISSUE #18638: Tutorial DataModule leads to Validation data leakage in Multiprocessing
### 📚 Documentation

It is recommended in the docs to do splits in the setup() method & [an example is given](https://pytorch-lightning.readthedocs.io/en/1.7.7/data/datamodule.html?highlight=random_split#what-is-a-datamodule) that does random splits in setup.

Perhaps it shouldn't be recommended to do random splits in the setup(). Because 
I've found that when I do random_split inside setup with multiprocessing it results in each GPU worker having a different validation_dataset (given by random split).

image

_So when training the model is trained on some (or all) of the validation data when random splits are done in setup()._ (without a seed)

cc @borda @carmocca @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

The `seed_everything` function can be called without arguments. In this case, it will take the seed value from the environment variable `PL_GLOBAL_SEED`. But if the environment variable is not set, it falls back to selecting the seed randomly. This isn't usually a problem if Lightning is in control of launching the processes, because child processes will inherit the seed value from the environment variable. However, in cases where processes are launched externally (e.g. SLURM, torchelastic etc.), this default behavior is dangerous because each process will independently choose a random seed. This can affect sampling, randomized validation splits, and other behaviors that rely on each process having the same seed. 

This PR removes this random behavior and makes it so that the `seed_everything` falls back to setting a fixed seed 0. This change only affects users who have intentionally called `seed_everything` with no arguments, and those who use `seed_everything=True` in LightningCLI. Those uses will now have to select the number randomly first, and then pass it in to `seed_everything`, unless they are happy with the new default of 0. 

Fixes #18638
(to be confirmed)

----
:books: Documentation preview :books:: https://pytorch-lightning--18846.org.readthedocs.build/en/18846/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Review by Borda — APPROVED** [review]
How about 42?

**Review by carmocca — APPROVED** [review]
This could be considered a breaking change for scripts that relied on this behaviour to run ablations

**Comment by awaelchli** [comment]
@carmocca I added this to the changed section already. Is there anything else I should do to communicate this change? A warning that the feature changed would be possible to add but then we are again offending all these users who don't want to see false-positive warnings.

**Comment by carmocca** [comment]
Nothing I can think of. I just wanted to point it out

File: src/lightning/fabric/utilities/seed.py [hunk 1]
```diff
@@ -36,20 +36,20 @@ def seed_everything(seed: Optional[int] = None, workers: bool = False) -> int:
...
             except ValueError:
-                seed = _select_seed_randomly(min_seed_value, max_seed_value)
+                seed = 0
                 rank_zero_warn(f"Invalid seed found: {repr(env_seed)}, seed set to {seed}")
     elif not isinstance(seed, int):
         seed = int(seed)
 
     if not (min_seed_value <= seed <= max_seed_value):
         rank_zero_warn(f"{seed} is not in bounds, numpy accepts from {min_seed_value} to {max_seed_value}")
-        seed = _select_seed_randomly(min_seed_value, max_seed_value)
```
Comment (jonathanking) [inline src/lightning/fabric/utilities/seed.py:52]: I would like to propose that this single line change be reverted and a ValueError raised instead of a warning. Instead of assuming that the user would prefer to use a random seed or a 0 when the provided seed is **invalid,** I think it's safer to mimic `numpy`'s behavior and raise a ValueError. I had been passing in improper seeds to pytorch lightning which were ignored (though I did miss the warning message). This meant that they were 0 when I expected them to be unique. A crash would have been preferred.

Comment (bhimrazy) [reply src/lightning/fabric/utilities/seed.py:52]: Thanks for bringing this up, @jonathanking. Since this PR is merged and quite old, maybe you could open an issue referencing it, so the team and community can discuss it further there? It would help get more eyes on the topic. 🙌

Best,
Bhimraj

Comment (deependujha) [reply src/lightning/fabric/utilities/seed.py:52]: Hi @jonathanking,

Thanks for sharing your perspective. The current behavior is intentional: defaulting to `0`, experiments remain deterministic while still warning the user about invalid seeds.

This avoids breaking existing workflows and aligns with `seed_everything()`’s role as a convenience helper.

Since the warning is already in place and documented, we don’t plan to change this behavior, as doing so **`could introduce breaking changes for existing users`**.

cc: @bhimrazy

Comment (deependujha) [reply src/lightning/fabric/utilities/seed.py:52]: and in case, you haven't read the PR description, please go through it.

Comment (jonathanking) [reply src/lightning/fabric/utilities/seed.py:52]: Thank you @deependujha  and @bhimrazy for your feedback! I will open a new issue for discussion.

@deependujha Yes, I did have a chance to read the PR description, and found that it is actually somewhat incorrect. 

> This change only affects users who have intentionally called seed_everything with no arguments, and those who use seed_everything=True in LightningCLI

The above is not true, this change affected users who pass in arguments to this function as well.
